napari/utils/migrations.py

```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
     _OldClass.__new__.__signature__ = prealloc_signature  # type: ignore [attr-defined]
 
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
+    """A dictionary that issues warning messages when deprecated keys are accessed.
+
+    Deprecated keys and values are not stored as part of the dictionary, so will not
+    appear when iterating over this or its items.
+
+    Instead deprecated items can only be accessed using `__getitem__`, `__setitem__`,
+    and `__delitem__`, or using `self.deprecations` directly.
+    """
+
+    # Maps from a deprecated key to its value and deprecation message.
+    deprecations: dict[str, tuple[Any, str]]
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.deprecations = {}
+
+    def __getitem__(self, key: str) -> Any:
+        if key in self.deprecations:
+            value, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            return value
+        return super().__getitem__(key)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if key in self.deprecations:
+            _, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            self.deprecations[key] = value, message
+            return None
+        return super().__setitem__(key, value)
+
+    def __delitem__(self, key: str) -> None:
+        if key in self.deprecations:
+            _, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            del self.deprecations[key]
+            return None
+        return super().__delitem__(key)
```

Comment:
Maybe `del` should also delete base key? 

I could imagine that deleting is to not pass given value to the layer constructor, and deleting deprecated properties without base will not be good.

Reply:
I think it's fine as is. The user gets a warning telling them something is wrong and we try our best to support the deprecated behavior.

Reply:
Unless we only support deprecations as renames, I think it will be tricky to implement something that will work well generally here. Another option is to raise with the deprecation message, which will break plugins that do this instead of warning them and continuing without effect (as now).

Reply:
As I understand, our only usecase is rename of parameter?

Reply:
There are other use cases in #5477